app/Decsys/Services/WebhookService.cs

```diff
@@ -21,5 +26,71 @@ IWebhookRepository webhooks
     /// <returns>The created Webhook</returns>
     public int Create(WebhookModel model)
         => _webhooks.Create(model);
+
+    
+    /// <summary>
+    /// Triggers webhooks from a given payload.
+    /// </summary>
+    /// <param name="payload">The payload to trigger and Post.</param>
+    public async Task Trigger(PayloadModel payload)
+    {
+        var webhooks = _webhooks.List(payload.SurveyId);
+        
+        foreach (var webhook in webhooks)
+        {
+            if (await FilterCriteria(webhook, payload))
+            {
+                await SendWebhook(payload, webhook.CallbackUrl);
+            }
+        }
+    }
+
+    /// <summary>
+    /// Filters the webhooks based on matching criteria.
+    /// Currently focussed on matching page numbers on page navigation.
+    /// </summary>
+    /// <param name="payload">Payload to check</param>
+    /// <param name="webhook">Webhook to match against.</param>
+    /// <returns>True if filter matches</returns>
+    private static Task<bool> FilterCriteria(WebhookModel webhook, PayloadModel payload)
+    {
+        foreach (var eventType in webhook.TriggerCriteria.SelectMany(filter => filter.EventTypes))
+        {
+            // check 1: if they are the same event types
+            var result = eventType.GetType() == payload.EventType.GetType();
+
+            if(result)
+                // check 2: type specific validation criteria
+                // add further criteria for further types
+                result = eventType switch {
+                    PageNavigation nav => CheckIsValid(nav, payload.EventType as PageNavigation),
+                    _ => false
+                };
+
+            return Task.FromResult(result);
```

Comment:
quick q. Does this even need to be async?

Reply:
If it does, shouldn't you add `async` to the method signature?

Reply:
Aha, my question isn't really "Does it need to be async?" but more "Does it need to return a Task if it's not?"

If it was async it would need the keyword, yes, and it would also 99% of the time return a Task.

But given it isn't async, is there value in it returning `Task<T>` instead of just `T`?

It'd save you from all the `Task.FromResult()` calls too.

Reply:
It really does not! Have updated